Warbands & Treaties

this-friday opened the pull request:
## About The Pull Request

_A variable antagonist operating from the Duchy's outskirts._

_Relations between members of the Warband are fragile, as its Lieutenants have a 50% chance of spawning as **Aspirants.**_

_The Warband's goal is to end the week with a signed Treaty._

---

<details>

<summary>WARBANDS</summary>


# Each Warband has the following,

- ### 1 <ins>Warlord</ins>

In the lobby phase, they choose the warband's theme and any modifiers that might interest them.
The availability of certain options will change based on the influence of certain Patrons & Storyteller(s).
Afterwards (and as long as they can keep their Lieutenants in check), they're considered the leader.

- ### 3 <ins>Lieutenants</ins>

Among which, any number of them can be Aspirants.
Aspirant status isn’t a secret. Warlords and other Lieutenants can recognize Aspirants at a glance. 
This should encourage both parties to play smart and make compromises with each other.

Additionally, Lieutenants can opt to openly desert the Warband.
During a desertion, they split off into a new warband with a fraction of their progenitor warband's strength.


- ### 2 <ins>Grunts to a Lieutenant</ins>

Poor gear, but strength-wise are generally equivalent to wretches.
As a side note, all Warband members are given the “Formation Fighter” trait. This allows them to tileswap through friendly NPCs

- ### <ins>Envoys</ins>

A temporary spawn from a Warlord or Lieutenant interacting with one of their recruitment points - at which point they swap characters and directly control said Envoy. 
They can be used as disposable proxies for diplomacy or schemes.
<ins>The Warcamp's exit is decided when an envoy uses their SCOUT A PATH verb,</ins> and envoys are <ins>initially the only characters capable of leaving.</ins>
There's a roll to determine if an Envoy is Immediately Killed by Shadow People. Tread carefully.


### All stats are flat and uninfluenced by races, virtues or statpacks. They get what they get.

</details>

<details>

<summary>GOONS & THE GOON-ASSOCIATED</summary>

# ATTENTION GOONERS
Warlords and Lieutenants may spawn squads of NPC Goons from the recruitment points in their camp.
Squad size varies by role. Regular Lieutenants have a squad size of 2. Aspirants and Warlords have a squad size of 4.
Classes might further tweak the size of a squad. The Rival Lord (Standard Warlord) has a squad size of 8. Every officer in a Peasant Rebellion has a squad size of 8, regardless of their role.

Commands vary based on your target.

**(Basic Commands)**
FOLLOW: Target yourself.
CHARGE: Target the world.
NEUTRAL: Target the Combat Mode button. Note that any command given or attack delivered against them will disrupt this.

**(Temp Commands) (16 minute cooldown)**
KILL: Target the STRONG INTENT button. (12 second STR buff)
SURVIVE: Target the DEFEND INTENT button. (20 second CON & WIL buff)

A charging Goon will move towards the tile you marked, and then attack anything unmarked as an ally nearby. Anything. They’ll attack maids. They’ll attack spectators. This Shit Means Nothing To Them, Man
The greater implications and nuances of the Hyperwar are lost on them. All they can do is Kill.

**If you want someone to remain unharmed, you’ll need to either…**
- Hold your NPCs in check by keeping them on Follow.
- Set your NPCs to Neutral.
- Tell them to clear out.
- Associate them.


Goons exposed to sewer or ocean tiles will immediately drown.

</details>

<details>

<summary>TREATIES</summary>

# Treaties

A Warband’s main objective is fulfilled through the signing of a treaty.

Currently, terms may be signed to:
#### Create or Remove Laws
- Required Signatures: Duke or Marshal
- Laws created by a treaty cannot be removed.

#### Freeze Laws
- Required Signature: The Duke
- No laws may be declared or purged.

#### Adjust Taxes
- Required Signatures: The Duke (or the Bishop, if the Church’s tax rate is targeted).
- The chosen tax group is adjusted, then frozen for the rest of the round.

#### Exile
- Required Signatures: The Target Themselves, the Bishop, or the Duke
- Requires 2 signatures.
- If an exiled character returns to the city - and if they step under the sun - they’ll catch fire.

#### Tribute
- Required Signatures: Target Faction Owner
- Exchange mammon between vaults.

#### Freeform
- Required Signatures: Whomever
- just type some shit, bro

#### Cede Territory
- Required Signatures: Target Faction Owner
- Cede a territory from one faction to another. Details can be found in the Territories section.

**Additionally, Warbands can offer unique terms. At the moment, the only unique term is given to the Sorcerer-King.**
#### Acknowledge Superior Wizard
- Required Signatures: Court Magician
- The Court Magician is emotionally crippled for the foreseeable future.


Once all terms on a treaty are signed, burning it at dawn will activate it.
Or, alternatively, it may be activated at any time by casting Sacred Flame on it

Warbands can spawn as many as they need by interacting with their Campaign Planner.
Councillors spawn with one.
Outside of this, they don’t exist.


</details>

<details>

<summary>TERRITORIES</summary>

# Territories

Territories/Treaty Factions are flavor + a convoluted offshore Meister account.
Each faction is given a vault. They may spend coin from that vault and import a crate of Prized Goods from one of their territories.


## Who gets them?
A handful of default territories are given out between:
### The Duke
	KINGSFIELD
		PRIZED GOOD: Arms (mostly iron, with a low chance for steel and an even lower chance to potentially import silver)
		Upon Tarichea's outskirts lies an untouched fragment of the Old City, left desolate after the hour of the Sundering.
		Centuries later, within its decayed marble halls and upon its overgrowth-choked bulkheads, a new strain of nobility plays their hand at civilization.

	THE AZURE PASS
		PRIZED GOOD: Tea
		Just beyond the city, simple folk ply their trades in quiet villages. For now, peace shrouds them from Psydonia's ever-brewing madness.

	THE TERRORBOG
		PRIZED GOOD: Spider Honey
		To be proclaimed Warden of the Terrorbog is a condemnation. It's land, yes - but it's a burden. Cursed earth.

### The Bishop
	TARICHEA
		PRIZED GOOD: Gems
		The capital of Old Azuria remains as naught but a shattered monument to Progress.
		All that remains is known as Mount Decapitation: a wound carved into one of the Empire's many necks, from which
		lava and forgotten treasure alike spill into condemned, unholy lands.
		By the former Bishop's decree - and no doubt with all Gods willing - condemned it shall remain.

	THE TOMB OF PSYVALUS
		PRIZED GOOD: Gold Trinkets
		Upon one of the diocese's lesser peaks lies an Astratan monastery.
		Its monks follow an orderly routine as ritual: brew wine, pray skyward, and commit suicide come nitefall.
		Upon sunrise, a select few return carrying revelations stolen from beyond death.
		The rest remain in the cold below, and pilgrims adorn their graves with gold.

### Heartfelt
	HEARTFELT
		PRIZED GOOD: Rocks. They’re also given two coins to import said rocks with.
		Heartfelt's misfortune is immortalized in its landscape.
		And as the Rot takes its toll on the rest of the continent, it shall soon find itself in good company.

### The Soilsons
	THE AZURE HIGHLANDS
		PRIZED GOOD: Grain
		A tract of grand farmland, nested upon the crest of the steep cliffs surrounding the southern bay.

### The Merchant
	NOC'S WAY
		PRIZED GOOD: Randomized
		A trade route cutting through a notably treacherous constellation in the Sea of Stars.
		The poisoned waters of Noc's Way glitter with a bluish-silver radiance, staining the hulls of passing ships and crippling the flesh of those who grace it.
		While once monopolized by a Lirvassi trading company, wise business left a notable share in the Merchant's hands.


### The Guildmaster
	THE BLACK ROAD
		PRIZED GOOD: Randomized (Limited to Materials)


### Prized Goods
Coin may be spent to spawn a *Prized Good Import* chest.
The chest is filled with a collection of items that are of an approximate value to the expected import.

**By the importer’s choice, it’ll either spawn:**
At the Docks with a 60% tax (bumped up to 90% if a Warband takes the Blockade aspect)
In the Grove with a 0% tax (bumped up to 30% if a Warband takes the Blockade aspect)

(side note: the tax doesn’t actually go anywhere, since it’s less of an actual Tax and more of an ‘import cost’. that shit's gone)

(side side note: each prized import has an item limit of 80. After that point, excess value spawns as silver coins.

(side side side note: the final value of an imported chest is divided by the territory’s ‘distance’ variable. This is irrelevant for the default territories, as they all have a distance of 1. Warband territories have a randomized distance between 1 and 3.

(side side side side note: a bandit wearing an unsilenced Rontz Ring will be alerted during a Grove Imports)

## How do you make a Prized Import?
### For the Duchy
Click the Prized Import section in the Nervemaster.

### Anyone Else
1. Get a signet ring.
2. Get tallow & a tallow pot
3. Get paper.
4. Fill the pot, melt the tallow, tallow the ring.
5. Use the tallowed signet ring on the paper.
6. Fill out the options (territory, value, location)
7. Feed the resulting Import Writ into a HERMES.

## How do you get money into a Territory Faction’s vault?
### For the Duchy
It’s directly connected to the treasury.

### Everyone Else

1. Get a signet ring.
2. Get tallow & a tallow pot
3. Get paper.
4. Get coin(s).
5. Use the coins on the paper.
6. Once you’re finished adding coins, use a tallowed signet ring on the final product.
7. Feed the resulting Grant into a HERMES.

<sub>(in short:
coin -> paper
signet ring -> coin-filled paper (now a grant)
grant -> HERMES)</sub>


**Alternatively, trade from another faction's vault via a Treaty.**


Additionally: 
A Grant can be attached to an existing import writ - at which point the import will attempt to use money from the grant, rather than the faction’s vault.
A grant may also be attached to a blank sheet of paper to create an import writ.
For convenience’s sake, Crownstones also function as tallowed signet rings. Don't ask how that works.


## How do you get the money back out?
### For the Duchy
It’s directly connected to the treasury.

### For Anyone Else
Via Prized Good imports.


## ON ECONOMIC CATASTROPHE
There are fairly large sums of money in the vaults of pre-generated factions (around 9200 between them all at roundstart).
It's fairly difficult for these factions to - on their own - actually translate that money 1 to 1 into real coins.

But since the Duchy's faction is directly connected to its vault, it's fairly easy for **them**.

So as a warning: the Duchy could achieve Mega Rich Light-Bending levels of mammon. On paper, it's connected to roleplay (extorting territory owners, making trades, etc.) rather than farming busywork, so maybe it'll be fun.


</details>

<details>

<summary>SPELLS, VERBS, TRAITS & OTHER DETAILS</summary>

## SPELLS
	Sweep (Given to Warlords and Lieutenants)
		A reskinned Repulse that only functions while surrounded by 3 or more enemies.
		(because honestly if you getting surrounded/worked down by Two Dudes that’s just your problem)
		Requires a held weapon.
		For every character swept beyond the third, its cooldown is reduced by 7 seconds and it will deliver an additional 15 points of damage.

	Associate (Given to Warlords and Lieutenants)
		Used to declare a character as an ally.
		Antagonists declared as allies increase the Disorder score of the Warband.
		Exiled characters cannot be re-associated as allies, save by the Warlord.
		If someone can’t be declared as a full ally, they’ll still receive your personal faction tag.
	
	Exile (Given to Warlords)
		Exiles a character from the Warband - or failing that, from your personal faction.
		Exiling an antagonist will clear the disorder they were contributing.
		If an exiled character is re-associated and exiled again a permanent stack of disorder is incurred.

## VERBS
		Communicate (Given to Everyone)
			Send a message to every member of the Warband.
			Requires the sender to be outside and for the Warband to have less than 8 disorder.
			The Warlord themselves can always use Communicate, regardless of the current disorder score.

		Shortcut (Given to Everyone)
			Upon reaching any travel or far travel tile, using Shortcut will take you back to the Warcamp.
			Requires a path to be scouted (Scout a Path verb), and for the Shortcut tile in the camp to be active (as it can potentially deactivated by an enemy interacting with it).

		Desert Warband (Given to Lieutenants)
			A Lieutenant may opt to Desert the warband and create their own.
			Aspirants Lieutenants steal a base percentage of 30
			Regular Lieutenants steal a base percentage of 3
			Each point of disorder increases the percentage of stolen troops by 15.
			Any subordinates or allies they associated are instead considered associated with their new warband.
			An exiled Lieutenant will automatically desert.

		Abandon Envoy (Given to Envoys)
			Kills the Envoy you’re controlling and sends you back to your original character.
			There’s potentially a style option for this suicide, which at the moment is possessed by two warband types (Ascendant Sect & Sorcerer-King).

		Scout a Path (Given to Envoys)*
			Use this verb near a Rock Wall someplace away from the city.
			The travel tiles leading to your warcamp will be spawned, along with a small outskirts map spawned as an intermission between the warcamp and the main map.
			The outskirts will vary depending on the region you scouted the path from.
			*During a Desertion, this verb is given to the deserting Lieutenant. It’ll spawn them a new Warcamp if there’s space.
			If there’s no space, it’ll just spawn a Recruitment Tile.

## TRAITS
		Formation Fighter
			Allows a character to tileswap with allied NPCs at the cost of stamina.
			The tileswap cannot be performed if they’ve recently been Feinted or Baited.

		Old Dog
			Prevents a character from gaining XP.
			Given to all Warband characters.

		Eye for Detail
			Allows for the interpretation and creation of Treaties.
			Given to Warlords and Lieutenants
			Townside, it is given to…
				The Duke and all Non-Combat Nobility
				The Priest, and Ravox/Astratan Acolytes
				The Inquisitor 
				The Guildmaster
				The Archivist
				Vagabond Scholars & Couriers
				Trader (Scholar)
				The Merchant
				Courtesans

			Exceptions among the above:
				An Inbred Lord, as his birthright demands.
				Knight-Commander Marshal is an exception if they’re any younger than Old. They still possess the authority to sign terms, but they are assumed to be too new to their position to have the same level of expertise as their councilors.
				Hand, with the exception of Advisor.
				Ordinator.

# OTHER
**Warning Letters**
When a Warband spawns, a random character is chosen to receive a warning letter. It contains hints about their aspects, subtype and main warband type.
Those most likely to receive the warning are a Hand, Councillor, Inquisitor, Marshal, Towner or the Soilson.
There is however, a 25% chance that the warning goes to someone entirely random.

If the Warband takes the Surprise aspect, there will be no warning.

**Baseturf Stuff**
Ports https://github.com/tgstation/tgstation/pull/88794 and most of /tg/’s baseturf code to help with template generation. And by “ports #88794”, I mean I just dragged the entire baseturfs.dm file from there to here.
(and then I didn’t actually delete the old baseturf procs because I was afraid to touch the generation for everything else. the github bot watched that happen and still decided to label the PR ’’’’code maintenance’’’’. This is effectively the Mandate of Heaven)


</details>


<details>

<summary>DISCLAIMER & TO-DO LIST</summary>

# DISCLAIMER

this is my first coding project, so i don’t know the best way to do things and a good majority of this code is from Copy Pasting
if you see a flub or something that can be improved please let me know
(especially with TGUI. I had No Clue what the Fuck was going on in there, even with help)

**with that in mind and compounded with the fact that I’m taking a break before I continue working on this, expect this to go very slowly.**

this is up as a draft in case I give up entirely (as i’m prone to) or vanish mysteriously (as i’m also prone to)

### TO-DO LIST
#### The Simple Ones

- [x] figure out why there’s no sunlight in the warcamps. it’s dark as Fuck in here Holy Shit
- [x] lock XP gain/skill farming from warband characters
- [x] make one (1) npc loadout for each warband
- [x] add an NPC encounter in the warband’s outskirts
- [x] ~~pretty up UI & warcamps a little bit. or maybe not.~~ Who Cares Bro


#### The Big Ones

- [x] finish up treaties
- [x] get a simplified Minion Order to work on complex mobs 
- [x] smooth out the lobby phase so the lieutenants (and maybe 1-2 grunts) spawn at roundstart with the warlord, rather than being summoned via rally point 
- [ ] testing


</details>

---

## Testing Evidence

https://i.imgur.com/MohIsxI.gif
👉👈 there'll be more when it's done
## Why It's Good For The Game

a majority of the current antag roster is Ontologically Evil™. thus, they themselves or their opposing party (the town) are inclined to be uncompromising at risk of shattering immersion

warbands are (in most cases) secular. they’re just Guys who - with the aspirant roll taken into account - probably don’t even like each other

includes the world’s silliest, most unbalanced classes. there’s a 20 perception plate archer in here. Who Gives A Fuck Anymore,

this-friday committed:
Warbands & Treaties

Ryan180602 commented:
Apart from the map issues. I just want to ask to be cleared. Was this made with AI? Or atleast with the help of it?

this-friday commented:
> Apart from the map issues. I just want to ask to be cleared. Was this made with AI? Or atleast with the help of it?
<img width="472" height="592" alt="345" src="https://github.com/user-attachments/assets/5800ff36-9715-4f86-ac56-abc177add4f9" />

a lot of the .tsx was, but in my defense google straight up just does them for you when you try searching a question
(i also can't be fucked to learn it right now. i'm of Weak Will)

this-friday committed:
Erm,

this-friday committed:
Erm, (2)

this-friday committed:
Erm, (3)

this-friday committed:
uuugh...,,,,, guhh,,,,,,,,u uuuugh........

WeNeedMorePhoron reviewed (commented):
Echoing what other devs said about AI usage AND a first time project.

this-friday committed:
dumping a placeholder track

JohnnyJohnJohnson commented:
Hi there, would ya like to sign my petition?

this-friday committed:
Divorce & Alimony (Psydonian Edition)

this-friday committed:
NPC Squads (Getting Stuck on Ledges Edition)

Ryan180602 commented:
Sooo- has anything come of this?

this-friday committed:
Treaties & Lobbies (Q2 2027 Edition)

this-friday committed:
Merge branch 'main' into warbandtest

this-friday committed:
Faithlocks + Conflict Fixes

this-friday committed:
by god it's cooked

this-friday committed:
by god it's cooked (2)

this-friday committed:
BY GOD IT'S COOKED (3)

this-friday committed:
ZELP

this-friday committed:
ZEEEELP

this-friday committed:
the cache-ening

- puts almost everything in caches. most importantly the Goon NPCs, since otherwise we freeze the entire game whenever we want to spawn a massive group
- throttles the AI & pathfinding of said goons. (they’re throttled enough that we can run 80 mob gmod style npc battles and only get like a +5-7% hit to time dilation, so it’s probably in a good place considering there isn’t gonna be anywhere near 80)
- finalizing the loadouts for Sects
- finalizing outskirts encounters
- finalizing the lobby chat
- sweep cooldown reduced to match Repulse (40 second cooldown down to 25 seconds)
- adds the Enlightenment verb for the Prophet (Sect Warlord)
- gives Warlords a unique warhorn option for the express purpose of aura farming (-5 Mood and a temporary combat music swap for whoever hears it).

this-friday committed:
fixing up marches

- fixes up long range pathfinding & throttling for marches
- makes spawn_objective() actually account for walls, so the cave outskirts map is no longer Cooked
- don't even look at what happened in observer.dm . it was a different time ok

this-friday committed:
faction ownership mini-refactor + proc organization

- faction ownership relies on job .type rather than a string
- splits the main warband manager's procs across 6 labeled files

this-friday committed:
casus belli(s)

- adds a "casus belli" stage to warband creation, in which the lobby members can propose, vote & choose the initial term/concept for their treaty
- should speed up brainstorming & lead to less deliberation once they actually spawn
- also adds a couple comments clarifying some of the 1-line changes

this-friday committed:
Merge branch 'main' into warbandtest

this-friday committed:
initial parity fixes

this-friday committed:
revokes the mandate of heaven (someone else ported the baseturf stuff from /tg/ except they did it properly)

this-friday committed:
parity fixes + descope prep

- parity fixes (converting everything mob-related to the new AI system & getting it working)
- descope prep (preparing to drop Territories, as they were handled more in-depth in another PR)
